### PR TITLE
yarn2nix: use sha512 by default

### PR DESCRIPTION
**Pull request by prusnak**
###### Motivation for this change

- this PR hopefully helps with https://github.com/NixOS/nixpkgs/issues/77238 for yarn packages
- if yarn.lock contains sha1 hash, we don't upgrade that to sha512, but we keep that one (maybe something to address in future PRs?)

###### Things done

- Built on platform(s)
  - [ ] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [x] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [22.05 Release Notes (or backporting 21.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2205-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
  - [ ] (Release notes changes) Ran `nixos/doc/manual/md-to-db.sh` to update generated release notes
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/development/tools/yarn2nix-moretea/yarn2nix/lib/fixPkgAddMissingSha512.js [hunk 1]
```diff
@@ -34,7 +34,7 @@ function getSha1(url) {
 }
 
 // Object -> Object
-async function fixPkgAddMissingSha1(pkg) {
+async function fixPkgAddMissingSha512(pkg) {
   // local dependency
```
Comment (jtojnar) [inline pkgs/development/tools/yarn2nix-moretea/yarn2nix/lib/fixPkgAddMissingSha512.js:39]: I would also add something like

```
if (pkg.integrity) {
  return pkg;
}
```

as according to my understanding, `integrity` has priority over the hash in `resolved`.

File: pkgs/development/tools/yarn2nix-moretea/yarn2nix/lib/fixPkgAddMissingSha512.js [hunk 2]
```diff
@@ -46,21 +46,21 @@ async function fixPkgAddMissingSha1(pkg) {
     return pkg
   }
 
-  const [url, sha1] = pkg.resolved.split('#', 2)
+  const [url, sha512] = pkg.resolved.split('#', 2)
```
Comment (jtojnar) [inline pkgs/development/tools/yarn2nix-moretea/yarn2nix/lib/fixPkgAddMissingSha512.js:49]: As mentioned, I am not sure if `resolved` field can contain anything other than `sha1`. Will prolly need to scour https://github.com/yarnpkg/yarn

File: pkgs/development/tools/yarn2nix-moretea/yarn2nix/lib/fixPkgAddMissingSha512.js [hunk 3]
```diff
@@ -46,21 +46,21 @@ async function fixPkgAddMissingSha1(pkg) {
...
 
-  if (sha1 || url.startsWith('https://codeload.github.com')) {
+  if (sha512 || url.startsWith('https://codeload.github.com')) {
     return pkg
```
Comment (jtojnar) [inline pkgs/development/tools/yarn2nix-moretea/yarn2nix/lib/fixPkgAddMissingSha512.js:52]: We should probably do `getSha512` if the `sha1` is present.

**Review by jtojnar — CHANGES_REQUESTED** [review]

File: pkgs/applications/editors/uivonim/yarn.nix [hunk 4]
```diff
@@ -6,5719 +6,5719 @@
...
         url  = "https://registry.yarnpkg.com/to-readable-stream/-/to-readable-stream-1.0.0.tgz";
-        sha1 = "ce0aa0c2f3df6adf852efb404a783e77c0475771";
+        sha512 = "Iq25XBt6zD5npPhlLVXGFN3/gyR2/qODcKNNyTMd4vbm39HUaOiAM4PMq0eMVC/Tkxz+Zjdsc55g9yyz+Yq00Q==";
       };
     }
     {
       name = "to_regex_range___to_regex_range_2.1.1.tgz";
       path = fetchurl {
         name = "to_regex_range___to_regex_range_2.1.1.tgz";
         url  = "https://registry.yarnpkg.com/to-regex-range/-/to-regex-range-2.
```
Comment (jtojnar) [inline pkgs/applications/editors/uivonim/yarn.nix:5697]: Example of sha1

**Comment by yu-re-ka** [comment]
The resolved field always uses sha1, so some parts of this change don't make sense. With #119522 we should already use sha512 most of the time?

Please also consider and test the IFD-mode of yarn2nix where we have to use what's in the yarn.lock file.

**Comment by prusnak** [comment]
@jtojnar @yu-re-ka Feel free to take over this PR, unfortunately, I won't be able to pursue this further in the following days.

**Pull request closed without merging**